src/main/java/net/minecraftforge/common/util/SortedRegistry.java

```diff
@@ -0,0 +1,219 @@
+        var index = elements.indexOf(element);
+        if (index == -1) return ImmutableList.of();
+        return elements.subList(0, index);
+    }
+
+    /**
+     * Returns the ordered list of elements that appear after the specified one.
+     * If the element is not found, an empty list is returned.
+     * @param element The element to look up
+     * @return An immutable list of successors in ascending order
+     */
+    public ImmutableList<T> getAllAfter(T element)
+    {
+        var index = elements.indexOf(element);
+        if (index == -1) return ImmutableList.of();
+        return elements.subList(index + 1, elements.size());
+    }
+
+    /**
+     * Returns the ordered list of elements that appear between the specified ones.
+     * If the first element is null or not found, the registry is indexed from the start.
+     * If the last element is null or not found, the registry is indexed until the end.
+     * @param first The lower bound
+     * @param last The upper bound
+     * @return An immutable list of elements between the arguments in ascending order
+     */
+    public ImmutableList<T> getAllBetween(@Nullable T first, @Nullable T last)
+    {
+        var firstIndex = first != null ? elements.indexOf(first) : -1;
+        var lastIndex = last != null ? elements.indexOf(last) : elements.size();
+        if (firstIndex >= lastIndex) return ImmutableList.of();
+        return elements.subList(firstIndex + 1, lastIndex);
+    }
+
+    public static class Builder<T>
+    {
+        private final Class<T> type;
+        private final BiMap<ResourceLocation, T> entries = HashBiMap.create();
+        private final Multimap<ResourceLocation, ResourceLocation> edges = HashMultimap.create();
+
+        public Builder(Class<T> type)
+        {
+            this.type = type;
+        }
+
+        /**
+         * Adds an element to the sorted registry.
+         * @param element The element
+         * @param name The name of the element
+         * @param predecessors A list of elements that must appear before this element (String, ResourceName, and T are allowed)
+         * @param successors A list of elements that must appear after this element (String, ResourceName, and T are allowed)
+         * @param <T2> The type of the element
+         * @return The element
+         */
+        public synchronized <T2 extends T> T2 add(T2 element, ResourceLocation name, List<Object> predecessors, List<Object> successors)
+        {
+            Objects.requireNonNull(predecessors, "Predecessor list may not be null.");
+            Objects.requireNonNull(successors, "Successor list may not be null.");
+            if (entries.containsKey(name))
+                throw new IllegalArgumentException("An entry has already been registered with this name: " + name);
+            entries.put(name, element);
+            for(var other : predecessors)
+                edges.put(getName(other), name);
+            for(var other : successors)
```

Comment:
Formatting: space before the opening parentheses in a `for`-expression:
```suggestion
            for (var other : successors)
```